### PR TITLE
HBP-87. Merge latest changes from Denis

### PR DESCRIPTION
**Pull request by liadomide**
We should 
- review the code here, as much as possible. If something takes too long to review, mark it for the future
- make unit-tests pass
- move integration tests into "examples"
- in the end merge in master
- then deploy in Open Shift Jupyter Hub

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 VERSION = "1.0.0"
```
Comment (liadomide) [inline setup.py]: In preparation for the Jupyter Hub deployment, we should increase the version number to `1.0.1`.

File: tvb_multiscale/config.py [hunk 2]
```diff
@@ -26,42 +32,211 @@
...
+                       "cortex": cortex.Cortex.from_file(os.path.join(DEFAULT_SUBJECT_PATH,
+                                                                      DEFAULT_CORT_SURFACE_ZIP),
+                                                         region_mapping_file=os.path.join(DEFAULT_SUBJECT_PATH,
+                                                                                          DEFAULT_CORT_REGION_MAPPING_TXT))
+                       }
+except:
+    DEFAULT_SUBJECT = None
+
+
+class GenericConfig(object):
```
Comment (liadomide) [inline tvb_multiscale/config.py]: This is again deprecated code. I have tried to remove these part several times already, and is keep appearing (probably after some bad merge).

File: tvb_multiscale/config.py [hunk 3]
```diff
@@ -26,42 +32,211 @@
...
+        folder = self._folder_figs
+        safe_makedirs(folder)
+        return folder
+
+    @property
+    def FOLDER_TEMP(self):
+        return os.path.join(self._out_base, "temp")
+
+
+class FiguresConfig(object):
```
Comment (liadomide) [inline tvb_multiscale/config.py]: Same until the end of this file!
`FiguresConfig ` has been migrate to tvb-library (next to the plots)

File: tvb_multiscale/io/h5_reader.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import os
+from collections import OrderedDict
+
+import h5py
+
+from tvb_multiscale.config import initialize_logger
+from tvb_multiscale.io.h5_writer import H5Writer
+
+
+class H5Reader(object):
```
Comment (liadomide) [inline tvb_multiscale/io/h5_reader.py]: Why not use the existing interface `tvb.core.neocom.h5 load(path)`?

File: tvb_multiscale/io/h5_writer.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+from tvb_multiscale.config import CONFIGURED, initialize_logger
+from tvb_multiscale.io.datatypes_h5 import REGISTRY
+
+from tvb.core.neocom import h5
+from tvb.contrib.scripts.utils.log_error_utils import warning
+from tvb.contrib.scripts.utils.data_structures_utils import is_numeric, ensure_list
+from tvb.contrib.scripts.utils.file_utils import change_filename_or_overwrite
+
+
+class H5Writer(object):
```
Comment (liadomide) [inline tvb_multiscale/io/h5_writer.py]: Why not use the existent `tvb.core.neocom.h5: store_complete(DT, folder)` ?

File: tvb_nest/interfaces/models.py [hunk 6]
```diff
@@ -1,16 +1,25 @@
...
+from tvb_nest.config import CONFIGURED, initialize_logger
 from tvb_nest.interfaces.base import TVBNESTInterface
-from tvb_scripts.utils.log_error_utils import initialize_logger
 
+from tvb.simulator.models.reduced_wong_wang_exc_io import ReducedWongWangExcIO
 from tvb.simulator.models.reduced_wong_wang_exc_io_inh_i import ReducedWongWangExcIOInhI
-from tvb.simulator.models.wilson_cowan_constraint import WilsonCowan
-from tvb.simulator.models.generic_2d_oscillator_multiscale import Generic2dOscillator
+from tvb.simulator.models.wilson_cowan_constraint import WilsonCowan as WilsonCowanSimModel
+from tvb.simulator.models.generic_2d_oscillator_multiscale import Generic2dOscillator as Generic2dOscillatorSimModel
```
Comment (liadomide) [inline tvb_nest/interfaces/models.py:9]: Whouldn't it be better if the Cosimulator sublaccs together with these `models.* multiscale ` classes would sit separately of tvb-library ?
Either under `tvb-contrib` or under `tvb-multiscale.core` ?

File: tvb_nest/tests/test_models.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+class TestIzhikevichRedWWexcIO(TestModel):
+    tvb_to_nest_mode = "rate"
+
+    def __init__(self):
+        super(TestIzhikevichRedWWexcIO, self).__init__(ReducedWongWangExcIO, list(range(10)),
+                                                       BasalGangliaIzhikevichBuilder, IzhikevichRedWWexcIOBuilder, {})
+
+
+@pytest.mark.skip(reason="These tests are taking too much time")
```
Comment (liadomide) [inline tvb_nest/tests/test_models.py]: Have you tried to run these with a length = 10ms, nest_populations_order=10 and transient=1, for example?

File: docker/Jenkinsfile [hunk 8]
```diff
@@ -11,7 +11,7 @@ pipeline {
             steps {
                 script {
                     LAST_SHA_ROOT = sh (script: "curl -s 'https://api.github.com/repos/the-virtual-brain/tvb-root/commits?sha=tvb-multiscale-cosim' | grep sha | head -1 | tr '\"' '_'", returnStdout: true).trim()
-                    LAST_SHA_MULTISCALE = sh (script: "curl -s 'https://api.github.com/repos/the-virtual-brain/tvb-multiscale/commits' | grep sha | head -1 | tr '\"' '_'", returnStdout: true).trim()
+                    LAST_SHA_MULTISCALE = sh (script: "curl -s 'https://api.github.com/repos/the-virtual-brain/tvb-multiscale/commits?sha=ongoing_dp_work' | grep sha | head -1 | tr '\"' '_'", returnStdout: true).trim()
```
Comment (liadomide) [inline docker/Jenkinsfile]: This will need to be reverted right before merge into master ?

File: docker/Dockerfile [hunk 9]
```diff
@@ -94,7 +94,7 @@ RUN $VENV/bin/pip3 install setuptools; \
 
 # Install TVB specific packages
 WORKDIR $HOME/packages
-RUN $VENV/bin/pip3 install pandas xarray jupyterlab pytest pytest-cov pytest-benchmark oct2py scikit-learn scipy; \
+RUN $VENV/bin/pip3 install pandas xarray jupyterlab pytest pytest-cov pytest-benchmark oct2py scikit-learn scipy elephant sqlalchemy flask; \
```
Comment (liadomide) [inline docker/Dockerfile]: Why flask?

**Review by liadomide — COMMENTED** [review]
Partial code review:
- why do we need separate top modules `tvb_multiscale`, `tvb_nest`, `tvb_utils`, and now even `tvb_elephant` ?
  If we were to put all these under `tvb_multiscale`, `tvb_multiscale.core, tvb_multiscale.nest, tvb_multiscale.utils, tvb_multiscale.elephant` it would be straight forward to have a Pypi release and to document it efficiently
- we should consider moving the cosimulator subclass (together with the multiscale models) under `tvb-contrib.multiscale` and outside of `tvb-library`

**Comment by liadomide** [comment]
@bvalean does this branch also need a rebase with master ?

**Comment by liadomide** [comment]
The top README should be revised: tvb-scripts no longer exists, we could write about the dependency relations between the other modules. 

If we merge all these into a single tvb_multiscale top module, we should directly produce a Pypi package on Jenkins

File: docker/Jenkinsfile (hunk 8 above)
Comment (bogdanvalean1) [reply docker/Jenkinsfile]: Yes.

File: tvb_nest/tests/test_models.py (hunk 7 above)
Comment (dionperd) [reply tvb_nest/tests/test_models.py]: This test is no longer slow. 
The problem was the simulation of a model SpikingReducedWongWang that has several modes (4th dimension) representing neurons.
I have excluded this model now, and also reduced the numbers as you are suggesting.

File: tvb_nest/interfaces/models.py (hunk 6 above)
Comment (dionperd) [reply tvb_nest/interfaces/models.py:9]: Indeed the multiscale models could be moved to either tvb-contrib or tvb-multiscale.core.tvb.models.
Whether the CoSimulator will also move there will be I guess a decision to make after we decide how to implement the TVB infrastructure for co-simulation, and how much of it should be in the scientific_library of TVB.

**Comment by dionperd** [comment]
> Partial code review:
> 
> * why do we need separate top modules `tvb_multiscale`, `tvb_nest`, `tvb_utils`, and now even `tvb_elephant` ?
>   If we were to put all these under `tvb_multiscale`, `tvb_multiscale.core, tvb_multiscale.nest, tvb_multiscale.utils, tvb_multiscale.elephant` it would be straight forward to have a Pypi release and to document it efficiently
> * we should consider moving the cosimulator subclass (together with the multiscale models) under `tvb-contrib.multiscale` and outside of `tvb-library`

We have moved towards a tvb_multiscale.core, .tvb_nest structure now.
However, I would insist to a tvb_nest name as opposed to nest, because tvb_nest is really a specification of tvb_multiscale, i.e., being an interface, it makes sense to have both names there.

Indeed, tvb_utils or tvb_elephant could become utils and elephant!

**Pull request merged**